moxin-protocol/src/open_ai.rs

```diff
@@ -0,0 +1,133 @@
+use std::collections::HashMap;
+use serde::{Deserialize, Serialize};
+use crate::data::*;
+
+// Based on https://platform.openai.com/docs/api-reference/chat/object
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct ChatRequestData {
+    // Not really necesary but it is part of the OpenAI API. We are going to send the id
+    // of the model currently loaded.
+    pub model: ModelID,
+
+    pub frequency_penalty: Option<f32>,
+    pub logprobs: Option<bool>,
+    pub top_logprobs: Option<u32>,
+    pub max_tokens: Option<u32>,
+    pub presence_penalty: Option<f32>,
+    pub seed: Option<u32>,
+    pub stop: Option<Vec<String>>,
+    pub stream: Option<bool>,
+    pub temperature: Option<f32>,
+    pub top_p: Option<f32>,
+
+    // Adding the following fields since there are part of the OpenAI API,
+    // but are not likely to be used in the first version of the client
+    pub n: Option<u32>,
+    pub logit_bias: Option<HashMap<String, f32>>,
+}
+
+// Shared structs for ChatResponse and ChatResponseChunk
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct MessageData {
+    pub content: String,
+    pub role: String,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct TopLogProbsItemData {
+    pub token: String,
+    pub logprob: f32,
+    pub bytes: Option<Vec<u8>>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct LogProbsItemData {
+    pub token: String,
+    pub logprob: f32,
+    pub bytes: Option<Vec<u8>>,
+    pub top_logprobs: Vec<TopLogProbsItemData>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct LogProbsData {
+    pub content: Vec<LogProbsItemData>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum StopReason {
+    #[serde(rename = "stop")]
+    Stop,
+    #[serde(rename = "length")]
+    Length,
+    #[serde(rename = "content_filter")]
+    ContentFilter
+}
```

Comment:
Could we merge ChatError into StopReason, as they essentially represent the same thing?

Reply:
Yes! `ChatError` is removed now. Stop reason is embedded inside the response struct (based on OpenAI format)